internal/directives/simple_engine_promote.go

```diff
@@ -77,94 +81,129 @@ func (e *SimpleEngine) executeSteps(
-			}, err
+				Status:                kargoapi.PromotionPhaseErrored,
+				CurrentStep:           i,
+				StepExecutionMetadata: stepExecMetas,
+				State:                 state,
+				HealthCheckSteps:      healthChecks,
+			}, fmt.Errorf("error getting runner for step %d: %w", i, err)
 		}
 
-		// Check if the step has exceeded the maximum number of attempts.
-		maxAttempts := step.GetMaxAttempts(reg.Runner)
-		if maxAttempts > 0 && attempt >= maxAttempts {
-			return PromotionResult{
-				Status:      kargoapi.PromotionPhaseErrored,
-				CurrentStep: i,
-				State:       state,
-				Attempt:     attempt,
-			}, fmt.Errorf("step %q exceeded max attempts", step.Alias)
+		// If we don't have metadata for this step yet, create it.
+		if int64(len(stepExecMetas)) == i {
+			stepExecMetas = append(stepExecMetas, kargoapi.StepExecutionMetadata{
+				Alias:     step.Alias,
+				StartedAt: ptr.To(metav1.Now()),
+			})
 		}
+		stepExecMeta := &stepExecMetas[i]
 
-		// Count the attempt we are about to make.
-		attempt++
-
-		// Execute the step.
+		// Execute the step
 		result, err := e.executeStep(ctx, promoCtx, step, reg, workDir, state)
+		if err != nil {
+			// Let a hard error take precedence over the result status and message.
+			stepExecMeta.Status = kargoapi.PromotionPhaseErrored
+			stepExecMeta.Message = err.Error()
+		} else {
+			stepExecMeta.Status = result.Status
+			stepExecMeta.Message = result.Message
+		}
+		state[step.Alias] = result.Output
 
-		// If the step failed, and the maximum number of attempts has not been
-		// reached, we are still "Running" the step and will retry it.
-		if err != nil || result.Status == kargoapi.PromotionPhaseErrored || result.Status == kargoapi.PromotionPhaseFailed {
-			if maxAttempts < 0 || attempt < maxAttempts {
-				var message strings.Builder
-				_, _ = message.WriteString(fmt.Sprintf("step %q failed (attempt %d)", step.Alias, attempt))
-				if result.Message != "" {
-					_, _ = message.WriteString(": ")
-					_, _ = message.WriteString(result.Message)
-				}
-				if err != nil {
-					_, _ = message.WriteString(": ")
-					_, _ = message.WriteString(err.Error())
-				}
-
-				// Update the result to indicate that the step is still running.
-				result.Status = kargoapi.PromotionPhaseRunning
-				result.Message = message.String()
-
-				// Swallow the error if the step failed, as we are still
-				// retrying it.
-				err = nil
+		if stepExecMeta.Status == kargoapi.PromotionPhaseSucceeded {
+			stepExecMeta.FinishedAt = ptr.To(metav1.Now())
+			stepExecMeta.ErrorCount = 0
+			if healthCheck := result.HealthCheckStep; healthCheck != nil {
+				healthChecks = append(healthChecks, *healthCheck)
 			}
+			continue // Move on to the next step
 		}
 
-		// Update the state with the step output, regardless of the result.
-		state[step.Alias] = result.Output
+		// Treat errors and logical failures the same for now.
+		// TODO(krancour): These may be handled differently in the future.
```

Comment:
I think what you are aiming at here is that we eventually may have terminating errors etc. But I am not sure we will remember this ourselves in 1-2 months, so I would add a "how" here.

Reply:
```suggestion
		// TODO(krancour): In the future, we should fail without retry for a logical failures and unrecoverable errors and retry only those with a chance of recovery.
```